### PR TITLE
Profile Pages with User Activity and PRs

### PR DESCRIPTION
**Pull request by GODrums**
### Motivation

Instead of redirecting to Github profiles, we want to provide a quick overview over a user's activity and stats (in the future also achievements). 

### Description

This PR adds a dynamic profile page for every user to the path `/user/{login}`:
- New data endpoint calculating all the required data server-sided `GET /user/{login}/profile`
- New Angular page `/user/{login}` with clickable elements for redirects to the specific PR/Review on Github
- Leaderboard entries now redirect to the user page instead of the Github profile

My initial Figma design prototype can be found [here](https://www.figma.com/design/3z7Ldlx4gqhqP4vBcVjazY/Hephaestus-Page-Designs?node-id=0-1&t=1ZCxCcrReNnpNhWm-1). 

### Testing
@milesha, would love to receive some feedback on my design choices as well! My inspiration was [Duolingo's website design](https://static.wikia.nocookie.net/duolingo/images/f/f0/Bandicam_2022-03-26_12-25-22-180.jpg) with a bit of my own flair.

### Screenshots (if applicable)
Screenshot 2024-10-08 230427
Screenshot 2024-10-08 230439

#### General

- [x] PR title is clear and descriptive
- [x] PR description explains the purpose and changes
- [x] Code follows project coding standards
- [x] Self-review of the code has been done
- [x] Changes have been tested locally
- [x] Screenshots have been attached (if applicable)
- [ ] Documentation has been updated (if applicable)

#### Client (if applicable)

- [x] UI changes look good on all screen sizes and browsers
- [x] No console errors or warnings
- [x] User experience and accessibility have been tested
- [x] Added Storybook stories for new components
- [x] Components follow design system guidelines (if applicable)

#### Server (if applicable)

- [x] Code is performant and follows best practices
- [x] No security vulnerabilities introduced
- [x] Proper error handling has been implemented
- [ ] Added tests for new functionality
- [ ] Changes have been tested in different environments (if applicable)

### CODE REVIEW

**Review by milesha — COMMENTED** [review]
Feedback on design choices 😁:
Overall I like the styling! 
Some stuff to consider:
1) Lay-outing of the upper part (icon, name, link, contribution date) aligned on the left and not in the center. Then you also can put the icons of the projects the person is contributing to to the right. I understand the idea of centring everything - but it's more of a mobile than web-practice, especially if there is much info around (in our case 4 other lines)
2) On the screenshot (not tested on my machine yet) the activities are not sorted based on the time (9 hours/20 days/a day/2 months ago) - it would however make total sense to have this list starting the the latest ones and anding with the oldest ones

File: webapp/src/app/home/profile/profile.component.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      return dayjs(this.query.data()?.firstContribution).format('Do [of] MMMM YYYY');
+    }
+    return '';
+  });
+
+  constructor(private route: ActivatedRoute) {
+    this.userLogin = this.route.snapshot.paramMap.get('id');
+  }
+
+  getRepositoryImage = (name: string) => {
```
Comment (milesha) [inline webapp/src/app/home/profile/profile.component.ts:78]: With a relatively short list of options (projects) it looks fine, but if we are adding more of those in the future (Iris, ...) then a mapping object could work a bit easier (up to you)

`const repoImages: { [key: string]: string } = {
  'Hephaestus': 'https://github.com/ls1intum/Hephaestus/raw/refs/heads/develop/docs/images/hammer.svg',
  'Artemis': 'https://artemis.in.tum.de/public/images/logo.png',
  'Athena': 'https://raw.githubusercontent.com/ls1intum/Athena/develop/playground/public/logo.png'
};

getRepositoryImage = (name: string) => repoImages[name.split('/')[1]] || 'https://avatars.githubusercontent.com/u/11064260?v=4';
`

**Comment by GODrums** [comment]
> Feedback on design choices 😁: Overall I like the styling! Some stuff to consider:
> 
> 1. Lay-outing of the upper part (icon, name, link, contribution date) aligned on the left and not in the center. Then you also can put the icons of the projects the person is contributing to to the right. I understand the idea of centring everything - but it's more of a mobile than web-practice, especially if there is much info around (in our case 4 other lines)
> 2. On the screenshot (not tested on my machine yet) the activities are not sorted based on the time (9 hours/20 days/a day/2 months ago) - it would however make total sense to have this list starting the the latest ones and anding with the oldest ones

Thank you for the feedback!

1. Agreed, my original plan was to left align it and put future gamification elements on the right hand side (for example trophies for leaderboard placement). Since those don't exist yet, centering felt like the best alternative to avoid huge blank spaces. The initial inspiration came from a Duolingo concept: 
image

2. Yes, the sorting was indeed bugged when I took that screenshot. It has been fixed with commit 4f2846a. Additionally, Felix suggested to limit the activity to the last 7 days (since the heading states "latest activity"), which is also already part of the PR (commit cde8aee)

File: server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequest.java [hunk 2]
```diff
@@ -78,4 +78,12 @@ public void addComment(IssueComment comment) {
     public void addReview(PullRequestReview review) {
         reviews.add(review);
     }
+
+    public Integer getAdditions() {
```
Comment (FelixTJDietrich) [inline server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequest.java:82]: Isn't this already generated by lombok?

File: webapp/src/app/home/profile/profile.component.html
Comment (FelixTJDietrich) [inline webapp/src/app/home/profile/profile.component.html:1]: Please create a story for this page :) You can just do it like the admin page etc., i.e. the story depends on the running server.

File: server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequest.java (hunk 2 above)
Comment (GODrums) [reply server/application-server/src/main/java/de/tum/in/www1/hephaestus/codereview/pullrequest/PullRequest.java:82]: At least my compiler required an explicit definition for it to work. I assume it has to do with the fact that the variables are of the primitive type `int`, while the methods require an `Integer` as return value.

Feel free to try if it works without them for you though.

File: webapp/src/app/home/profile/profile.component.html
Comment (GODrums) [reply webapp/src/app/home/profile/profile.component.html:1]: Resolved as part of #119.

**Review by FelixTJDietrich — APPROVED** [review]
LGTM, thank you for this PR! I think it turned out great :)

**Pull request merged**